### PR TITLE
fix: add support for protobuf 5.x

### LINKED ISSUE #628: Support protobuf 5.26.0 when released
Protobuf is releasing a new version [5.26.0](https://pypi.org/project/protobuf/#history) with a breaking change. It will [check the timestamp values](https://github.com/protocolbuffers/protobuf/blob/d2b71b347df6c8af5c3ee1c833f41b7772e3f250/python/google/protobuf/internal/well_known_types.py#L214) and raise an error for out of bound values. 

Apache Beam Python SDK depends on `google-api-core` that has dependency of `protobuf<5`. Apache Beam will prefer to upgrade to this new version of protobuf whenever it is out. 

What are the plans/timeline to release new version of `google-api-core` with support for[ protobuf 5.26.0](https://pypi.org/project/protobuf/#history)?

### LINKED ISSUE #641: Add pre-release testing
There is a `prerelease_deps` nox session for packages in `google-cloud-python`, [for example](https://github.com/googleapis/google-cloud-python/blob/9f0608987b8b72f9e613b170b91b6ab40493287f/packages/google-cloud-access-approval/noxfile.py#L340), that tests against pre-release versions of direct dependencies. We should create a similar presubmit check in this repository that tests pre-release versions of dependencies. For example, when a new pre-release version of `protobuf` is published, there should be a prerelease presubmit check in this repo that run tests against it.

### LINKED ISSUE #642: `mypy` presubmit fails with `types-protobuf==5.x`
The `mypy` [nox session](https://github.com/googleapis/python-api-core/blob/828ffe1d7b2edd36be5e852524adb7f5d1241770/noxfile.py#L175) fails with the [latest version](https://pypi.org/project/types-protobuf/5.26.0.20240422/) of `types-protobuf==5.26.0.20240422` and `types-protobuf==5.26.0.20240420`. This is related to [protobuf 5.x](https://pypi.org/project/protobuf/#history) which includes a breaking change compared with 4.x where `including_default_value_fields` is removed in `MessageToDict` : https://github.com/protocolbuffers/protobuf/commit/26995798757fbfef5cf6648610848e389db1fecf

See the stack trace below and [build log](https://github.com/googleapis/python-api-core/actions/runs/8859767950/job/24329889797?pr=634) from PR https://github.com/googleapis/python-api-core/pull/634
```
Run nox -s mypy
nox > Running session mypy
nox > Creating virtual environment (virtualenv) using python3.10 in .nox/mypy
nox > python -m pip install '.[grpc]' mypy
nox > python -m pip install types-setuptools types-requests 'types-protobuf!=4.24.0.20240106' types-mock types-dataclasses
nox > mypy google tests
mypy.ini: [mypy]: python_version: Python 3.7 is not supported (must be 3.8 or higher)
google/api_core/operations_v1/transports/rest.py:187: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:200: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:268: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:281: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:342: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:355: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:4[14](https://github.com/googleapis/python-api-core/actions/runs/8859767950/job/24329889797?pr=634#step:5:15): error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:424: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
google/api_core/operations_v1/transports/rest.py:436: error: Unexpected keyword argument "including_default_value_fields" for "MessageToDict"  [call-arg]
.nox/mypy/lib/python3.10/site-packages/google-stubs/protobuf/json_format.pyi:23: note: "MessageToDict" defined here
Found 9 errors in 1 file (checked 97 source files)
nox > Command mypy google tests failed with exit code 1
nox > Session mypy failed.
Error: Process completed with exit code 1.
```

### PR DESCRIPTION
**Pull request by parthea**
Fixes https://github.com/googleapis/python-api-core/issues/628 🦕
Fixes https://github.com/googleapis/python-api-core/issues/641 🦕
Fixes https://github.com/googleapis/python-api-core/issues/642 🦕

Regarding the `including_default_value_fields` argument of `json_format.MessageToDict()` in `protobuf`:

In protobuf version `3.19.6` - The default value of `including_default_value_fields` was `False`
https://github.com/protocolbuffers/protobuf/blob/5cba162a5d93f8df786d828621019e03e50edb4f/python/google/protobuf/json_format.py#L92

At the time that the argument was removed in protobuf 5.x - the default value of `including_default_value_fields` was `False`
https://github.com/protocolbuffers/protobuf/commit/26995798757fbfef5cf6648610848e389db1fecf#diff-8de817c14d6a087981503c9aea38730b1b3e98f4e306db5ff9d525c7c304f234

IOW, setting argument `including_default_value_fields` to `False` had no effect as it was the default behaviour.

The `Unit tests / unit_prerelease-3.12` presubmit check was added to run tests against the latest pre-release version of `protobuf` as several dependencies have a constraint on `protobuf<5`

### CODE REVIEW

**Comment by anndro** [comment]
Hello @parthea is there any scheduled date for release this pull request?

**Comment by zhangskz** [comment]
@parthea Is there anything blocking this PR still?

File: noxfile.py
Comment (vchudnov-g) [inline noxfile.py:147]: This is more succinct and clearer in highlighting the difference between the two branches, IMO:

```suggestion
        session.install(
            "-e",
            ".[grpc]" if install_grpc else ".",
            "-c",
            f"{constraints_dir}/constraints-{session.python}.txt",
            )
```

File: noxfile.py
Comment (vchudnov-g) [inline noxfile.py:132]: I'm confused: in line 206, where we call `default(..., prerelease=True)`, the decorator constrains `python=PYTHON_VERSIONS[-1]`. But here we are referencing the constraints for  `PYTHON_VERSIONS[0]`. Could you clarify?

File: noxfile.py
Comment (vchudnov-g) [inline noxfile.py:79]: Sorry, I have gaps in my mental model for testing prerelease;

This installs our dependencies at pre-release versions, right?

Is the only difference, then, the use of `--pre` below? If it's the same set of dependencies, it seems it would be clearer to have them loaded in the same place for pre-release and stable versions, and add the extra `install` parameters conditionally on each thing we install.

(It might be easier to talk synchronously about this.)

File: noxfile.py
Comment (vchudnov-g) [inline noxfile.py:204]: So this tests running against the pre-release versions of the dependencies, right? And we only bother to do this with the latest Python run-time.

A comment might be helpful. Also, my personal preference would to rename this (and similar code elsewhere) as `unit_with_prerelease_deps`. Every time I read something like `unit_prerelease` my mind first assumes we're testing the pre-release version of _this_ library, not of its dependencies.

File: .github/workflows/unittest.yml
Comment (vchudnov-g) [inline .github/workflows/unittest.yml:14]: I think this more succinct code should work (I haven't tested it) and would be easier to maintain. 
Ref: https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations

```python
        option: ["", "_grpc_gcp", "_wo_grpc"]
        python:
          - "3.7"
          - "3.8"
          - "3.9"
          - "3.10"
          - "3.11"
          - "3.12"
        exclude:
          - option: "_wo_grpc"
            python: 3.7
          - option: "_wo_grpc"
            python: 3.8
          - option: "_wo_grpc"
            python: 3.9
         include:
          - option: "_prerelease"
            python: 3.12
```

**Review by vchudnov-g — COMMENTED** [review]
A couple of suggestions, and some questions, but nothing major. This looks good!

File: google/api_core/operations_v1/transports/rest.py
Comment (parthea) [reply google/api_core/operations_v1/transports/rest.py]: Fixed in https://github.com/googleapis/python-api-core/pull/644/commits/5a49a53603c97a6d65deb9b23bb084699ef3372a

File: noxfile.py
Comment (parthea) [reply noxfile.py:132]: As per the comment in below, the constraints file for the lowest supported python version contains a list of all of the dependencies of the library. The function `install_prerelease_dependencies` will extract the dependencies  from the file.

https://github.com/googleapis/python-api-core/blob/126b5c7154223da5f0537c90462a49e7e0d3ad97/testing/constraints-3.7.txt#L1-L7

https://github.com/googleapis/python-api-core/blob/7fbce0dfe60502f4b9e8502f2bef11dc9a2c551b/noxfile.py#L79-L88

Once we have a list of dependencies, we will install them independently with the `--pre` and `--no-deps` option. `--upgrade is also added to ensure that we get the latest pre-release.

https://github.com/googleapis/python-api-core/blob/7fbce0dfe60502f4b9e8502f2bef11dc9a2c551b/noxfile.py#L99-L100

File: noxfile.py
Comment (parthea) [reply noxfile.py:79]: There is a difference in that we don't allow transitive dependencies  in the `install_prerelease_dependencies` session as we want to ensure that we're installing the pre-release version of each dependency. In the non-prerelease session, we are installing transitive dependencies.

File: noxfile.py
Comment (parthea) [reply noxfile.py]: Fixed in https://github.com/googleapis/python-api-core/pull/644/commits/e1efa83cb43ddd9413e540fbdab90763a7231e87

File: .github/workflows/unittest.yml
Comment (parthea) [reply .github/workflows/unittest.yml]: This comment is obsolete with the changes in https://github.com/googleapis/python-api-core/pull/644/commits/e1efa83cb43ddd9413e540fbdab90763a7231e87

**Comment by parthea** [comment]
@ohmayr PTAL

File: noxfile.py
Comment (ohmayr) [inline noxfile.py:237]: Just to clarify, do we want to run this against all versions?

File: google/api_core/operations_v1/transports/rest.py
Comment (ohmayr) [inline google/api_core/operations_v1/transports/rest.py:204]: This is essentially also converting a message to a dict. So we could use a similar approach to what we're doing for `request_params` and define a helper function so we're consistent. But it's more of a preference and could be a follow up.

I'll leave it up to you since this isn't a blocker!

**Review by ohmayr — APPROVED** [review]

File: google/api_core/operations_v1/__init__.py
Comment (parthea) [reply google/api_core/operations_v1/__init__.py:19]: I ran `black` on all of the files and this file was formatted

File: noxfile.py
Comment (parthea) [reply noxfile.py:237]: We normally only run it on 1 version but we can expand it if there is value

**Comment by JackCloudman** [comment]
Hello, I am eagerly anticipating the next release! Could you please provide any updates on the release date? Thank the whole team for your hard work and dedication.

File: google/api_core/operations_v1/transports/rest.py
Comment (parthea) [reply google/api_core/operations_v1/transports/rest.py:204]: I took a closer look at making this change and I don't believe the helper function will result in fewer lines of code. The reason that we had the helper function elsewhere was to remove duplication for the protobuf 3.x/4.x compatibility code but we don't have the duplication here.

**Pull request merged**
